protocol_driver_grpc_async_callback.cc

```diff
@@ -316,4 +323,177 @@ void ProtocolDriverGrpcAsyncCallback::ShutdownServer() {
+};
+
+}  // anonymous namespace
+ProtocolDriverServerGrpcAsyncCq::ProtocolDriverServerGrpcAsyncCq()
+    : thread_pool_((absl::base_internal::NumCPUs() + 1) / 2) {}
+
+ProtocolDriverServerGrpcAsyncCq::~ProtocolDriverServerGrpcAsyncCq() {}
+
+absl::Status ProtocolDriverServerGrpcAsyncCq::InitializeServer(
+    const ProtocolDriverOptions& pd_opts, int* port) {
+  std::string netdev_name = pd_opts.netdev_name();
+  auto maybe_ip = IpAddressForDevice(netdev_name);
+  if (!maybe_ip.ok()) return maybe_ip.status();
+  server_ip_address_ = maybe_ip.value();
+  server_socket_address_ = SocketAddressForIp(server_ip_address_, *port);
+  traffic_async_service_ = absl::make_unique<TrafficAsyncServiceCq>();
+  grpc::ServerBuilder builder;
+  builder.SetMaxMessageSize(std::numeric_limits<int32_t>::max());
+  std::shared_ptr<grpc::ServerCredentials> server_creds =
+      MakeServerCredentials();
+  builder.AddListeningPort(server_socket_address_, server_creds, port);
+  builder.AddChannelArgument(GRPC_ARG_ALLOW_REUSEPORT, 0);
+  ApplyServerSettingsToGrpcBuilder(&builder, pd_opts);
+  builder.RegisterService(traffic_async_service_.get());
+  server_cq_ = builder.AddCompletionQueue();
+  server_ = builder.BuildAndStart();
+
+  server_port_ = *port;
+  server_socket_address_ = SocketAddressForIp(server_ip_address_, *port);
+  if (!server_) {
+    return absl::UnknownError("Grpc Traffic service failed to start");
+  }
+
+  LOG(INFO) << "Grpc Traffic server listening on " << server_socket_address_;
+  // Proceed to the server's main loop.
+  handle_rpcs_ = absl::make_unique<std::thread>(
+      &ProtocolDriverServerGrpcAsyncCq::HandleRpcs, this);
+  return absl::OkStatus();
+}
+
+void ProtocolDriverServerGrpcAsyncCq::SetHandler(
+    std::function<std::function<void()>(ServerRpcState* state)> handler) {
+  static_cast<TrafficAsyncServiceCq*>(traffic_async_service_.get())
+      ->SetHandler(handler);
+  handler_ = handler;
+}
+
+absl::StatusOr<std::string> ProtocolDriverServerGrpcAsyncCq::HandlePreConnect(
+    std::string_view remote_connection_info, int peer) {
+  ServerAddress addr;
+  addr.set_ip_address(server_ip_address_.ip());
+  addr.set_port(server_port_);
+  addr.set_socket_address(server_socket_address_);
+  std::string ret;
+  addr.AppendToString(&ret);
+  return ret;
+}
+
+void ProtocolDriverServerGrpcAsyncCq::HandleConnectFailure(
+    std::string_view local_connection_info) {}
+
+void ProtocolDriverServerGrpcAsyncCq::ShutdownServer() {
+  server_->Shutdown();
+  server_cq_->Shutdown();
+  if (handle_rpcs_->joinable()) {
+    handle_rpcs_->join();
+  }
+}
+
+std::vector<TransportStat>
+ProtocolDriverServerGrpcAsyncCq::GetTransportStats() {
+  return {};
+}
+
+void ProtocolDriverServerGrpcAsyncCq::HandleRpcs() {
+  new CallData(traffic_async_service_.get(), server_cq_.get(), &handler_,
+               &thread_pool_);
+  void* tag;
+  bool ok;
+  while (server_cq_->Next(&tag, &ok)) {
+    if (!ok) {
+      delete static_cast<CallData*>(tag);
```

Comment:
I think this may lead to problems if ProcessGenericRPC was called when we do that. 

To handle all case cleanly, we may need to call unref here and make sure the refcnt is right.

Instead of setting it at 2 from the beginning, we may start at 1 and increment when in ProcessGenericRPC

Reply:
Done